flake.nix

```diff
@@ -0,0 +1,91 @@
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-21.11";
+    flake-utils.url = "github:numtide/flake-utils";
+    concat = {
+      url = "github:con-kitty/concat/wavewave-flake";
+      inputs.nixpkgs.follows = "nixpkgs";
+      inputs.flake-utlis.follows = "flake-utils";
+    };
+  };
+  outputs = { self, nixpkgs, flake-utils, concat }:
+    flake-utils.lib.eachSystem [ "x86_64-linux" "x86_64-darwin" ] (system:
+      let
+        parseCabalProject = import ./parse-cabal-project.nix;
+        excluded = [
+          "categorifier-vec-integration"
+          "categorifier-vec-integration-test"
+        ];
+        categorifierPackages = let parsed = parseCabalProject ./cabal.project;
+        in builtins.filter ({ name, ... }: !(builtins.elem name excluded))
+        parsed;
+        categorifierPackageNames =
+          builtins.map ({ name, ... }: name) categorifierPackages;
+        haskellOverlay = self: super:
+          builtins.listToAttrs (builtins.map ({ name, path }: {
+            inherit name;
+            value = self.callCabal2nix name (./. + "/${path}") { };
+          }) categorifierPackages);
+
+      in {
+        # This package set is only useful for CI build test.
+        # In practice, users will create a development environment composed by overlays.
+        packages = let
+          packagesOnGHC = ghcVer:
+            let
+              overlayGHC = final: prev: {
+                haskellPackages = prev.haskell.packages.${ghcVer};
+              };
+
+              newPkgs = import nixpkgs {
+                overlays = [ overlayGHC (concat.overlay.${system}) ];
+                inherit system;
+                config.allowBroken = true;
+              };
+
+              newHaskellPackages = newPkgs.haskellPackages.override (old: {
+                overrides =
+                  newPkgs.lib.composeExtensions (old.overrides or (_: _: { }))
+                  haskellOverlay;
+              });
+            in builtins.listToAttrs (builtins.map ({ name, ... }: {
+              name = ghcVer + "_" + name;
+              value = builtins.getAttr name newHaskellPackages;
+            }) categorifierPackages);
+
+        in builtins.trace (builtins.elemAt categorifierPackageNames 0)
+        (packagesOnGHC "ghc8107" // packagesOnGHC "ghc884"
+          // packagesOnGHC "ghc901" // packagesOnGHC "ghc921");
+
+        # see these issues and discussions:
+        # - https://github.com/NixOS/nixpkgs/issues/16394
+        # - https://github.com/NixOS/nixpkgs/issues/25887
+        # - https://github.com/NixOS/nixpkgs/issues/26561
+        # - https://discourse.nixos.org/t/nix-haskell-development-2020/6170
+        overlay = final: prev: {
+          haskellPackages = prev.haskellPackages.override (old: {
+            overrides =
+              final.lib.composeExtensions (old.overrides or (_: _: { }))
+              haskellOverlay;
+          });
+        };
+
+        devShell = let
+          ghcVer = "ghc8107";
```

Comment:
I think the default for the shell should be at least 9.0.1, to allow for LinearTypes in development, but why not just make it the most recent? Especially once there’s ghc-debug out, tracking the latest seems best for dev.

Reply:
Ah. no obvious reason on my side. GHC 8.10.7 is the default in NixOS 21.11, so I chose that. 
In fact, this devShell should be extended to multiple dev shells on different versions of GHCs. :-) 
I strongly agree having GHC >= 9.0 as default to allow for the new features. will update this soon.

Reply:
made ghc921 be default dev shell.

Reply:
uh-oh. ghc 9 builds was not successful (it seems like testing for PyF failed.) reverted the change (ghc8107 -> ghc921) until it is resolved.

Reply:
Ah yeah, I fought with PyF for a bit ... it needs some precise version bounds. Should be able to pin it based on the range in cabal.project?

Reply:
Oh, I think we can just skip testing PyF ... like, we don't run tests for upstream packages in Cabal. The test failures all seem to be a simple `s/[Char]/String/` difference that’s just noise.

Reply:
now dontCheck PyF. :-)